spec/integration/api/transform_spec.rb

```diff
@@ -0,0 +1,247 @@
+      transformation: { name: base_names, other_attrs: { type: trans_types, tweak_source: trans_tweak_sources, template: trans_templates } },
+      template: { name: base_names, other_attrs: { type: temp_types, pattern: temp_patterns } },
+      alphabet: { name: base_names, other_attrs: { alphabet: alpha_sets } },
+    }
+
+    transform_types.each do |type, attrs|
+      # With a given type...
+      context "#{type.capitalize}" do
+        attr_values = attrs[:other_attrs].values
+        attr_keys = attrs[:other_attrs].keys
+
+        # Iterate over its possible input options to provide a full permutation of options
+        # i.e. if I have a type with 3 inputs that have 3 options each, this will provide all 27 permutations for testing 
+        (0...(attr_values[0]&.length || 1)).each do |index_0|
+          (0...(attr_values[1]&.length || 1)).each do |index_1|
+            (0...(attr_values[2]&.length || 1)).each do |index_2|
+              let(:opts) do
+                opts = {}
+                opts[attr_keys[0].to_sym] = attr_values[0][index_0] if attr_values[0]
+                opts[attr_keys[1].to_sym] = attr_values[1][index_1] if attr_values[1]
+                opts[attr_keys[2].to_sym] = attr_values[2][index_2] if attr_values[2]
+                opts
+              end
+              
+              before(:context) do
+                @name = "#{attrs[:name].call}_#{type}".downcase
+              end
+
+              context "CRUD", order: :defined do
+                describe "#create_#{type}" do
+                  it "creates a #{type} in the vault server" do
+                    subject.transform.send("create_#{type}", @name, **opts)
+                    expected_opts = opts.clone
+                    expected_opts.delete(:tweak_source) if opts[:tweak_source] == "internal"
+                    if type == :transformation
+                      expected_opts[:allowed_roles] = [] 
+                      expected_opts[:templates] = [opts[:template]]
+                    end
+
+                    expected = Vault::Transform.const_get(type.to_s.capitalize).new(expected_opts)
+                    expect(subject.transform.send("get_#{type}", @name)).to eq(expected)
+                  end
+                end
+
+                describe "##{type}s" do
+                  it "lists the names of all #{type}s present on the vault server" do
+                    expect(subject.transform.send("#{type}s")).to include(@name)
+                  end
+                end
+
+                describe "#delete_#{type}" do
+                  it "removes the #{type} from the server" do
+                    subject.transform.send("delete_#{type}", @name)
+                    expect(subject.transform.send("#{type}s")).not_to include(@name)
+                  end
+                end
+              end
+            end
+          end
+        end
+      end
+    end
+
+    describe "#encode and #decode" do
```

Comment:
Could also add a case for batch_input encode/decode.

Reply:
I'll go ahead and add a test for batch_input. Should I be putting a builtin limiter for that so that we ensure we don't even send requests with more than a given batch size?

Reply:
If you're testing with a real server, the http layer _should_ be protecting against that, but a test wouldn't hurt :)